buildsystem/modules/FindSDL2.cmake

```diff
@@ -132,6 +132,27 @@ if(SDL2_LIBRARY_TEMP)
 	mark_as_advanced(SDL2_LIBRARY_TEMP)
 
 	set(SDL2_FOUND TRUE)
+
+endif()
+
+if(SDL2_INCLUDE_DIR AND EXISTS "${SDL2_INCLUDE_DIR}/SDL_version.h")
+	#to avoid extensive parsing of source files, compile a small program
+	#which returns the needed version string
+	try_run(SDL_VERSION_RUN
+		SDL_VERSION_TEST
+		"${CMAKE_BINARY_DIR}"
+		SOURCES "${CMAKE_CURRENT_LIST_DIR}/FindSDL_version.cpp"
+		LINK_LIBRARIES ${SDL2_LIBRARY}
+		CMAKE_FLAGS "-DINCLUDE_DIRECTORIES=${SDL2_INCLUDE_DIR}" "-DCMAKE_CXX_STANDARD=17"
+		RUN_OUTPUT_VARIABLE SDL2_VERSION_OUTPUT
+	)
+
+	if(SDL_VERSION_TEST)
+		message(STATUS "SDL versions: ${SDL2_VERSION_OUTPUT}")
+		set(SDL2_VERSION_STRING ${SDL2_VERSION_OUTPUT})
+	else()
+		message(FATAL "uhh something failed")
```

Comment:
That's not nice for cross-compilation. Why can't it use sdl-provided define?

Reply:
I don't follow, what define do you mean and how can we improve the patch? :)

Reply:
In cross-compilation you wouldn't usually be able to execute programs, but you should be able to get SDL version (both compiled against and linked to in runtime) using the API exposed in https://github.com/SDL-mirror/SDL/blob/e50629c5a126c82180ccbeb298f4e2360fa3925f/include/SDL_version.h#L60

Reply:
@janisozaur Could you please give an example on how you would do it?

Reply:
Sorry for the delay :)
I get what you mean, but then we're back to what the pull request originally did: It parsed the header file to extract the values with a regex.

Then I suggested using the built in API function to fetch the version because it seems like the official way to get it. But I understand that for cross-compilation this wouldn't work.

Can't we build the version-fetching-miniprogram with "`HOSTCC`" and then run it? CMake surely supports something like that, and we'd still get the version through the API.

Reply:
My point still holds: why do you need it in the cmake in the first place? Surely not just for the sake of reporting the patch level of SDL version found during build time, as that has no value to the person compiling… so what for?

Reply:
We have it in the build system so it can generate strings that contain the version of libraries the game was linked to. And this has to be done at compile time (hence CMake) to store the compiletime linkage versions.
What's your suggestion on how we should do it instead? Parse the header file (I mean if we have to, so be it, we can write helper functions that find a `#define` that can then also be used for other library headers)?

Reply:
But why? Why can't you just report the version from define?

Reply:
We extract the version at compile-time so it can be stored in `openage/config.py` and then printed in `openage/__init__.py` as `LONGVERSION`.
So, we store the versions at compile time to python.

The alternative, and that's what you're suggesting, right?, is to do this in C++, and just use the header file and get the value. It does seem like the better approach, and the python code could fetch C++ library versions via an libopenage-api-call and print python library versions directly. Then we'd have unified the approach and there would be no need for a program run at compiletime.

Implementing that is a bit more complicated, but seems the even better aproach, you're right :)

Reply:
Yes, just compile it in via the define from actual include and if need be extract via runtime query on target platform.

Reply:
@Mese96 Do you feel you can do it this way?

Reply:
Yeah, if this is now the way to go ;)
I need to look how to do "an libopenage-api-call" but otherwise it should be ok. 
If I understand this correctly this also should include a more general approach, useable for all version numbers ?

Reply:
yes correctly, this is a generic approach where we store the version numbers of C libraries in C++ instead of a python file (which we create through hacks like parsing the header file or compiletime-program-runs).

The approach is the following:
* store the version variables only in c++ in the config.h file (that method is already implemented)
* add a function that can read these version variables and return them as a key-value map
* access the version variables through cython: we could have a `versions.pyx` file that calls to the above function and converts the c++ key value map with all the versions to a python `dict`
* implement the python version printing by calling the cython function that fetches the versions from c++.

An example of the cython-c++ interface is e.g. our `util/path` system or the `compress/lxzd`. Have a look at:
* https://github.com/SFTtech/openage/blob/master/doc/code/pyinterface.md
* libopenage/util/compress/lzxd.h

If you need further help with cython, just ask me in the chatroom, please :)